src/inference/gibbs.jl

```diff
@@ -46,52 +53,46 @@ mutable struct GibbsState{V<:VarInfo, S<:Tuple{Vararg{Sampler}}} <: AbstractSamp
     samplers::S
 end
 
-function GibbsState(model::Model, samplers::S) where S<:Tuple{Vararg{Sampler}}
+function GibbsState(model::Model, samplers::Tuple{Vararg{Sampler}})
     return GibbsState(VarInfo(model), samplers)
 end
 
-const GibbsComponent = Union{Hamiltonian,MH,PG}
-
 function Sampler(alg::Gibbs, model::Model, s::Selector)
-    info = Dict{Symbol, Any}()
-
-    n_samplers = length(alg.algs)
-    samplers = Array{Sampler}(undef, n_samplers)
-    space = Set{Symbol}()
+    # sanity check for space
+    space = getspace(alg)
+    pvars = get_pvars(model)
+    @assert issubset(pvars, space) "[Gibbs] symbols specified to samplers ($space) doesn't cover the model parameters ($pvars)"
 
-    for i in 1:n_samplers
-        sub_alg = alg.algs[i]
-        if isa(sub_alg, GibbsComponent)
-            samplers[i] = Sampler(sub_alg, model, Selector(Symbol(typeof(sub_alg))))
-        else
-            @error("[Gibbs] Unsupported sampling algorithm $sub_alg")
-        end
-        space = (space..., getspace(sub_alg)...)
+    if !issetequal(pvars, space)
+        @warn("[Gibbs] extra parameters specified by samplers don't exist in model: $(setdiff(space, pvars))")
     end
 
-    # Sanity check for space
-    @assert issubset(get_pvars(model), space) "[Gibbs] symbols specified to samplers ($space) doesn't cover the model parameters ($(get_pvars(model)))"
-
-    if !(issetequal(get_pvars(model), space))
-        @warn("[Gibbs] extra parameters specified by samplers don't exist in model: $(setdiff(space, get_pvars(model)))")
+    # create tuple of samplers
+    samplers = let algs=alg.algs, model=model
```

Comment:
Is the `let` block here necessary?

Reply:
I guess not since the type of `samplers` can not be inferred in any case. In DiffEq we had some issues with captured variables in closures leading to performance regressions, and hence I sometimes just apply the [let-block trick](https://github.com/JuliaLang/julia/issues/15276#issuecomment-297596373), even though it might not be required.

Reply:
I rarely have a problem with closures these days, so I suspect this is overkill. Please check it with `@code_warntype`.

Reply:
Sure, I'll just remove it. I tried some simple examples (e.g., with `gdemo_default`), and in any case the type can't be inferred (it's just `Tuple`), so it doesn't seem to make a difference.